### PR TITLE
feat(perf-issues): update GroupSerializer et al. to support performance issues

### PR DESCRIPTION
**Pull request by barkbarkimashark**
To support Performance Issue Details, we need to update the various `GroupSerializers` to handle heterogeneous data appropriate for the group type. The current GroupSerializer assumes Issues are tightly coupled to errors. This PR attempts to separate the seams where we need to do issue type-specific stuff.

Things done so far:
- (ref) reorganize methods on `GroupSerializerBase`
- (ref) move StreamGroupSerializer*(list-based group serializers) into a separate file
- (feat) add feature flag `projects:performance-issue-details-backend` to expose this functionality to internal testing
- (feat) add `get_issue_type` and `get_issue_category` stub methods in `Group` model.
- (feat) split up retrieving `seen_stats` call to appropriately query the right place by Issue type.

Planning to put up another PR to implement the performance-specific snuba querying. 

Will hold off on merging to master until the `Group` columns are appropriately added.

### CODE REVIEW

File: src/sentry/api/serializers/models/group.py [hunk 1]
```diff
@@ -173,46 +172,267 @@ def __init__(
...
+            group_dict["isUnhandled"] = attrs["is_unhandled"]
+        if "times_seen" in attrs:
+            group_dict.update(self._convert_seen_stats(attrs))
+
+        # XXX(gilbert): remove this after performance issues internal testing
+        from sentry import features
+
+        if features.has("projects:performance-issue-details-backend", obj.project):
+            group_dict["issue_category"] = obj.get_issue_category()
+            group_dict["issue_type"] = obj.type
```
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/group.py:339]: Only expose these fields internally.

File: src/sentry/api/serializers/models/group.py [hunk 2]
```diff
@@ -173,46 +172,267 @@ def __init__(
...
+    def _seen_stats_performance(self, perf_issue_list, user) -> Mapping[Any, SeenStats]:
+        pass
 
-        return max(
-            min(last_seen - timedelta(days=1), datetime.now(pytz.utc) - timedelta(days=14)),
-            datetime.now(pytz.utc) - timedelta(days=90),
-        )
+    @abstractmethod
+    def _seen_stats_error(self, error_issue_list, user) -> Mapping[Any, SeenStats]:
+        pass
```
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/group.py:368]: Introduce type-aware methods for querying `seen_stats`.

File: src/sentry/api/serializers/models/group.py [hunk 3]
```diff
@@ -173,46 +172,267 @@ def __init__(
...
-        if seen_stats:
-            for item in seen_stats.values():
-                if last_seen is None or (item["last_seen"] and last_seen > item["last_seen"]):
-                    last_seen = item["last_seen"]
+        # bulk query for the seen_stats by type
+        error_stats = self._seen_stats_error(error_issues, user) or {}
+        perf_stats = (self._seen_stats_performance(perf_issues, user) if perf_issues else {}) or {}
+        agg_stats = {**error_stats, **perf_stats}
+        # combine results back
+        return {group: agg_stats.get(group, {}) for group in item_list}
```
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/group.py:360]: This seems to be the only performance-issue specific thing we need to update, besides the `issue_type` and `issue_category`

File: src/sentry/models/group.py [hunk 4]
```diff
@@ -563,6 +563,21 @@ def get_event_type(self):
...
+    def get_issue_category(self):
+        # TODO(gilbert): change this impl when category is available and mappable
+        if not self.get_issue_type:
+            return None
+
+        if self.get_issue_type in frozenset(["PERFORMANCE_N_PLUS_ONE", "PERFORMANCE_SLOW_SPAN"]):
+            return "performance"
+        else:
+            return "error"
+
```
Comment (barkbarkimashark) [inline src/sentry/models/group.py:580]: Introduce retrieving `issue_type` and `issue_category` here. Impl is currently stubbed out until the migration for `Group` can be merged in.

File: src/sentry/api/serializers/models/group.py [hunk 5]
```diff
@@ -164,7 +156,14 @@ class BaseGroupSerializerResponse(BaseGroupResponseOptional):
...
 
-class GroupSerializerBase(Serializer):
+class SeenStats(TypedDict):
+    times_seen: int
+    first_seen: datetime
+    last_seen: datetime
+    user_count: int
+
+
+class GroupSerializerBase(Serializer, ABC):
```
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/group.py:166]: This is now an `ABC`, any reason why it wasn't that before?

File: src/sentry/models/group.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/models/group.py:580]: I'm handling this in https://getsentry.atlassian.net/browse/WOR-2085 WIP here: https://github.com/getsentry/sentry/pull/37889

Comment (barkbarkimashark) [reply src/sentry/models/group.py:580]: Nice. I'll rebase this branch once we have your stuff merged and update my code to call yours properties.

File: tests/snuba/api/endpoints/test_group_details.py [hunk 6]
```diff
@@ -177,3 +177,24 @@ def test_assigned_to_unknown(self):
...
+        url = f"/api/0/issues/{group.id}/"
+
+        response = self.client.get(url, {"collapse": ["stats"]}, format="json")
+        assert response.status_code == 200
+        assert int(response.data["id"]) == event.group.id
+        assert response.data["stats"]
+        assert response.data["count"] is not None
+        assert response.data["userCount"] is not None
+        assert response.data["firstSeen"] is not None
+        assert response.data["lastSeen"] is not None
```
Comment (barkbarkimashark) [inline tests/snuba/api/endpoints/test_group_details.py:200]: Not sure if 'collapse' is actually used for 'stats' in the GroupDetails endpoint in practice. The collapse param isn't passed to the GroupSerializer* at all when we serialize a group - this isn't the case for the StreamGroupSerializer* which does actually collapse the stat fields.

Update this test if we ever decide to support collapsing fields in the GroupDetails endpoint.

File: src/sentry/models/group.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/models/group.py:594]: FYI this was merged, so you can rebase now and use it.

**Comment by barkbarkimashark** [comment]
Gonna split theses up into separate PRs for easier reviewing experience.

**Pull request closed without merging**